File: CCC_2018/J5/main.py
n = int(input())
paths = {}
min_steps = 99999999

# initialize all possible paths in a dictionary
for i in range(n):
    line = input().split(" ")
    m = int(line[0])
    connected = []
    for j in range(m):
        connected.append(int(line[j + 1]))
    paths[i + 1] = connected

# set all visited to false
visited = []
for i in range(n):
    visited.append(False)

queue = []
queue.append((1, 1))
visited[0] = True
while len(queue) > 0:
    node = queue.pop(0)
    current_page = node[0]
    steps = node[1]
    if len(paths[current_page]) == 0 and steps < min_steps:
        min_steps = steps
    else:
        for path in paths[current_page]:
            if not visited[path - 1]:
                queue.append((path, steps + 1))
                visited[path - 1] = True

all_visited = True
for visit in visited:
    if not visit:
        all_visited = False

print("Y" if all_visited else "N")
print(min_steps)

File: CCC_2018/J5/main.cpp
#include <bits/stdc++.h>
#include <queue>
using namespace std;

int n;
int paths[10001][10001];
bool visited[10001];
int min_steps = INT_MAX;
int main(){
    std::cin >> n;
    for (int i = 1; i <= n;i++) {
        int m;
        std::cin >> m;
        paths[i][0] = m;
        for (int j = 1; j <= m;j++) {
            std::cin >> paths[i][j];
        }
    }

    queue<std::pair<int, int>> queue;
    queue.push({1, 1});
    visited[1] = true;
    while (!queue.empty()) {
        std::pair<int, int>node = queue.front();
        queue.pop();
        int current_page = node.first;
        int steps = node.second;
        if ((paths[current_page][0] == 0) && (steps < min_steps)) {
            min_steps = steps;
        } else {
            for (int i = 1;i <= paths[current_page][0];i++) {
                if (!visited[paths[current_page][i]]) {
                    queue.push({paths[current_page][i], steps + 1});
                    visited[paths[current_page][i]] = true;
                }
            }
        }
    }

    bool all_visited = true;
    for (int i = 1;i <= n;i++) {
        if (!visited[i]) {
            all_visited = false;
        }
    }

    printf("%c\n%d\n", all_visited? 'Y':'N', min_steps);
    return 0;
}